### DISCUSSION TITLE
Any instructions on setting up for MacOS

### DISCUSSION THREAD

**Discussion by rongcuid** [lwjglgamedev/vulkanbook#59] [Q&A]
I am trying to make my code work on Mac, but there are some additional steps:

1. Detect if portability enumeration layer is present. If so, enable it and the portability enumeration flag
2. Allow portability devices

An additional issue is that LWJGL bundles MoltenVK by default, which does not contain any validation layer. I added `-Dorg.lwjgl.vulkan.libname=/opt/homebrew/Cellar/vulkan-loader/1.3.257/lib/libvulkan.1.3.257.dylib` to make it use system vulkan loader, but I still can't enable validation layer, getting errors like:

```
09:12:00.026 [ERROR] - [eng.graph.vk.Instance$Companion] - VkDebugUtilsCallback, dlopen(libVkLayer_khronos_validation.dylib, 0x0005): tried: 'libVkLayer_khronos_validation.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibVkLayer_khronos_validation.dylib' (no such file), '/Library/Java/JavaVirtualMachines/jdk-17.0.4.1.jdk/Contents/Home/bin/./libVkLayer_khronos_validation.dylib' (no such file), '/Library/Java/JavaVirtualMachines/jdk-17.0.4.1.jdk/Contents/Home/bin/../lib/libVkLayer_khronos_validation.dylib' (no such file), '/usr/lib/libVkLayer_khronos_validation.dylib' (no s
09:12:00.026 [ERROR] - [eng.graph.vk.Instance$Companion] - VkDebugUtilsCallback, Requested layer "VK_LAYER_KHRONOS_validation" failed to load!
Exception in thread "main" java.lang.RuntimeException: Error creating instance: -6
```

However, it seems like at least part of the validation is working... otherwise I wouldn't even see these message. I'd love to have some guideline of setting up on Mac

**Comment by lwjglgamedev** [comment]
Hi,

No Mac here, so just guessing, maybe you could try to add a symbolic link in any ofther dirs listed about to the library (wthout the version)?

**Reply by rongcuid** [reply]
I changed to the LunarG SDK and it seems to work. I suspect it is because homebrew installs two different versions of vulkan.

I suggest adding the handling of portability subset in the initialization code, though

**Comment by lwjglgamedev** [comment]
Glad to hear that. You need Lunar SDK also in other platforms to use validation layers. Could you please describe the missing steps you needed to perform in odrer t omake it work? I do not get what you mean by hombrew. If you can helpf me in describening the list of steps iI can include them

**Reply by rongcuid** [reply]
I have not finished converting yet (just finished porting the Instance, not device), and I am using Kotlin instead of Java. I also break down the instance creation process a bit more than the provided code, but I can show you the steps

**Reply by rongcuid** [reply]
Basically, if the instance supports `VK_KHR_portability_enumeration` (or `VK_KHR_PORTABILITY_ENUMERATION_EXTENSION_NAME`), you need to enable this extension as well as the flag `VK_INSTANCE_CREATE_ENUMERATE_PORTABILITY_BIT_KHR` during instance creation. Devices that don't have portability subset would not have this extension available.

**Reply by rongcuid** [reply]
I think that's all it requires

**Reply by rongcuid** [reply]
An additional thing is when creating device, if physical device contains `VK_KHR_portability_subset`, this extension must be enabled

**Comment by lwjglgamedev** [comment]
Can you check if this class fixes the issue with the instance Creation?
[Instance.zip](https://github.com/lwjglgamedev/vulkanbook/files/12136894/Instance.zip)

**Reply by lwjglgamedev** [reply]
Sorry, use this version instead, an please paste the value of os name logged.
[Instance.zip](https://github.com/lwjglgamedev/vulkanbook/files/12136897/Instance.zip)

**Comment by lwjglgamedev** [comment]
Just uploaded a fix-. If you see something that does not work, please fill a bug. Thanks for reporting this.
